### ISSUE TITLE
Can't resolve require

### ISSUE THREAD

**Issue by kaisellgren** [web-infra-dev/rspack#4989]
### System Info

Using the latest version: "@rspack/cli": "0.4.3"

### Details

Trying to compile NodeJS server code with:
```javascript
const path = require('node:path')

/**
 * @type {() => import('@rspack/cli').Configuration}
 */
module.exports = ({}, { mode }) => {
  return {
    mode,
    devtool: mode === 'development' ? 'nosources-cheap-module-source-map' : 'nosources-source-map',
    target: 'node18',
    context: __dirname,
    output: {
      publicPath: '/',
      filename: '[name].js',
      path: path.resolve(__dirname, 'dist/server'),
    },
    entry: {
      index: './src/server/server.ts',
    },
    optimization: {
      minimize: false,
    },
    module: {
      rules: [
        {
          test: /\.tsx?$/,
          exclude: [/node_modules/],
          loader: 'builtin:swc-loader',
          options: {
            sourceMap: true,
            jsc: {
              parser: {
                syntax: 'typescript',
                jsx: true,
                dynamicImport: true,
                decorators: true,
                topLevelAwait: true,
              },
              keepClassNames: true,
            },
          },
          type: 'javascript/auto',
        },
      ]
    },
    externalsType: 'node-commonjs',
    externalsPresets: {
      node: true,
    },
    watchOptions: {
      ignored: /node_modules/,
    },
  }
}

```

Produces:
> Can't resolve require ./event-listener-count in /home/xxx/node_modules/koa-router/node_modules/depd/lib/compat/index.js

This is the source code that it fails at:

```javascript
lazyProperty(module.exports, 'eventListenerCount', function eventListenerCount () {
  return EventEmitter.listenerCount || require('./event-listener-count')
})
```

It looks like the node module is dynamically requiring a dependency and Rspack for some reason is unable to follow this. 

### Reproduce Steps

Things worked fine in the earlier 0.3.6 version when I wasn't using the `builtin:swc-loader`. I upgraded to this newer 0.4.3 version and enabled this builtin loader and I'm not able to get this to function.

Also, this seems to only happen when passing `--production` mode flag, because I don't get these errors when doing a development build.

I've been looking for some **node**-specific flag I could turn on to maybe aid the builder, but I'm not finding any.

Edit: the problem actually goes away when enabling `newTreeShaking` option.

**Comment by JasonBoy** [comment]
same problem(ok in 0.4.2 with `builtin:swc-loader`), also not work in dev mode

**Comment by darkenergy96** [comment]
I faced this issue too immediately after upgrading from v0.3x to v0.4x for a react project.

![image](https://github.com/web-infra-dev/rspack/assets/16614251/876e703f-0544-48cc-ba92-3e1786610b06)
```js
// It's not able to probably resolve and bundle this require statement below
//  and the require statement is showing up in the browser and failing
module.exports = !require('./_descriptors') && !require('./_fails')(function () {
  return Object.defineProperty(require('./_dom-create')('div'), 'a', { get: function () { return 7; } }).a != 7;
});

```

**Comment by bvanjoi** [comment]
> module.exports = !require('./_descriptors') && !require('./_fails')(function () {
>  return Object.defineProperty(require('./_dom-create')('div'), 'a', { get: function () { return 7; } }).a != 7;
> });

I believe this issue had been fixed in #5000
